README.md
## What is 'thmigpen'?
Python 3.5 scripts to post \#ThemeMonday-related notifications and public channel messages to pnut.io.

### Scripts:
* **thmigpen** - Though probably inelegant, intended initially to be called by a cron job a week before a \#ThemeMonday event and run 1 to 3 times throughout the day at times to suit Far Eastern, European and North American audiences.  It's important to note it'll only post on one day a month.
* **thmigpoll** - A quite alpha rudimentary poll vote collector which checks a specific hashtag.  The script for e.g. `#tmpoll201803` will examine a predefined list of candidates, reject non-matching text and add matches to the total of votes.  Intended to be called by a cron job and run at frequent intervals over a small range of dates each month.
* **thmigask** - A very rudimentary poll suggestion collector which checks a specific hashtag.  The script for e.g. `#tmask201803` will check the hashtag, reject non-matching text and eventually compile and save a list of poll options. Intended to run before thmigpoll at frequent intervals over a small range of dates each month.


CHANGELOG.md
## Changelog
(most recent first)

### v0.1 2018-03-15:
* **`thmigpoll` only**: Added post numbers to votes as a precursor to reply confirmation of votes. Tidied messages, re-ordered some code.

### v0.1 2018-03-13:
* **`thmigpoll` only**: Added notification if no-one voted.

### v0.1 2018-03-12:
* **`thmigpoll` only**: Tidied up poll summary output.

### v0.1 2018-03-12:
* **`thmigpoll` only**: Trapped posts without a hash attached to the tag.

### v0.1 2018-03-11:
* **`thmigpoll` only**: Added vote totals.

### v0.1 2018-03-10:
* Added ancillary, very alpha, **`thmigpoll.py`** - a rudimentary script to list respondents to a poll; entirely dependent on an RSS feed of a pnut.io hashtag.  
* Edited `thmigpen` pnut channel code comments.

### v0.0 2018-03-07:
* First commit with basic functionality
 * Builds a list of only 1 year of first Mondays before the \#ThemeMonday second Mondays, starting 2018-01-01.


thmigask.py
#!/usr/bin/env python3

# thmigask v0.1.0 for Python 3.5

# A quite alpha attempt to list suggestions for a pnut.io poll, using the poll's suggestion hashtag.
# Based on thmigpoll.

# Setup tag and channel parameters, a list of valid poll options, and an empty list for votes: 
tag = 'tmask'
suggestionmade = False
retrievecount = 50
channelid = '962'

# Import @33MHz and @thrrgilag's library for interacting with pnut.io:
import pnutpy

# Setup pnut.io authorisation:
tokenfile = open("pnut_app_token.txt", "r")
token = tokenfile.read()
token = token.strip()
pnutpy.api.add_authorization_token(token)

# Get hashtag content from pnut.io:
d = pnutpy.api.posts_with_hashtag(tag, count = retrievecount)

# Extract posts, strip out unnecessary words, and construct a summary message:
posttext = 'The suggestions for the #' + tag + ' poll so far:\n'
number = retrievecount
hashtag = ''
while number >= 0:
	try:
		if not "is_deleted" in d[0][number]:
			user = str(d[0][number]["user"]["username"])
			suggestion = d[0][number]["content"]["text"]
			postnum = str(d[0][number]["id"])
			words = suggestion.split()
			# If '<=>' does not appear in a post it's not a notification, so process it:
			if not ('<=>' in suggestion):
				for word in words:
					if not ('#' in word):
						hashtag = 'nothing (no hashtag included, oops!) Please try again.'
					if ('#' in word):
						if word != ('#' + tag):
							suggestionmade = True
							hashtag = word
		if suggestionmade:
			# Create a suggestion entry:
			posttext += '• @' + user + ' suggested ' + hashtag + '\n'
	except IndexError:
		pass
	number -= 1

# Tidy the message if no-one voted yet:
if not suggestionmade:
	posttext += '• No-one has suggested a theme yet. Why not be first?! Simply create a public post using #' + tag + ' and your choice, not forgetting the hash.\n'

# FOR TESTING, uncomment the lines in this section to prevent the creation of posts & messages:
# print(posttext)
# posttext = ''

# If there's text to post:
if posttext:
	# Create message in channel 962, using ONLY the text from pnut_message:
	messagecontent = pnutpy.api.create_message(channelid, data={'text': posttext})
	
	# Create a public post:
	pollalert = 'To see the current suggestions for the TEST #' + tag + ' poll go to the Patter room link below. (Report not run automatically.)'
	channelurl = "https://patter.chat/room/" + channelid
	# Removed hash before tag:
	channelurlmd = '[' + tag + ' <=>](' + channelurl + ")"
	pollalert += "\n" + channelurlmd
	postcontent = pnutpy.api.create_post(data={'text': pollalert})


thmigpen.py
#!/usr/bin/env python3

# thmigpen
# v0.1 for Python 3.5

# Basis of the recurring repeat modified from: https://stackoverflow.com/questions/2295765/generating-recurring-dates-using-python

# Import date modules:
import dateutil.rrule as dr
import dateutil.parser as dp
import dateutil.relativedelta as drel
import datetime

# Import @33MHz and @thrrgilag's library for interacting with pnut.io:
import pnutpy

start=dp.parse("01/01/2018")   # First Monday in January 2018

# Setup a list of dates, adjust count to something larger, in multiples of 12 (momths):
rr = dr.rrule(dr.MONTHLY,byweekday=drel.MO(1),dtstart=start, count=12)

# What's today's date?:
t = datetime.date.today()
t = t.strftime('%d/%m/%Y')

x = ([d.strftime('%d/%m/%Y') for d in rr[::1]])

# If today is a week to go until #ThemeMonday:
pnut_message = ''
n = 0
while n <= 11:
	tmprev = (x[n])
	if tmprev == t:
		pnut_message = 'There\'s a week to go until #ThemeMonday. Let\'s suggest themes and start a poll before the weekend!'
	n += 1

# If a message gas been created:
if pnut_message != '':
	# Setup pnut.io authorisation:
	tokenfile = open("pnut_app_token.txt", "r")
	token = tokenfile.read()
	token = token.strip()
	pnutpy.api.add_authorization_token(token)

	# Create a public post using the text from pnut_message, add an x-post footer then create the post:
	posttext = pnut_message
	channelurl = "https://patter.chat/room/779"
	channelurlmd = "[#ThemeMonday <=>](" + channelurl + ")"
	posttext += "\n\n" + channelurlmd
	postcontent = pnutpy.api.create_post(data={'text': posttext})
	
	# Create message in channel 779, using ONLY the text from pnut_message:
	channelid = 779
	postcontent = pnutpy.api.create_message(channelid, data={'text': pnut_message})


thmigpoll.py
#!/usr/bin/env python3

# thmigpoll v0.1.12 for Python 3.5

# A quite alpha attempt to list respondents to a pnut.io poll, using the poll's hashtag.
# Based on rssupdatepnut and thmigpen.

# Setup tag and channel parameters, a list of valid poll options, and an empty list for votes: 
tag = 'tmpoll'
retrievecount = 50
channelid = '962'
polloptions = {
	'#test': 0,
	'#test1': 0,
	'#test2': 0
	}
votes = []

# Import @33MHz and @thrrgilag's library for interacting with pnut.io:
import pnutpy

# Import module to work out the maximum votes within the dict:
import operator

# Setup pnut.io authorisation:
tokenfile = open("pnut_app_token.txt", "r")
token = tokenfile.read()
token = token.strip()
pnutpy.api.add_authorization_token(token)

# Open vote post numbers file:
y = open('pollpostnumbers.txt', 'r')
y = y.readlines()

# List the poll options:
posttext = 'The choices in #' +  tag + ' are:\n'
for option in polloptions:
	posttext += '• ' + option + '\n'
posttext += '\n'

# Get hashtag content from pnut.io:
d = pnutpy.api.posts_with_hashtag(tag, count = retrievecount)

# Extract posts, strip out unnecessary words, check for matches to poll options, and construct a summary message:
votesmade = False
# Open the previous post numbers file:
f=open('pollpostnumbers.txt','w')
posttext += 'The votes so far:\n'
number = retrievecount
hashtag = ''
while number >= 0:
	try:
		if not "is_deleted" in d[0][number]:
			user = str(d[0][number]["user"]["username"])
			p_title = d[0][number]["content"]["text"]
			postnum = str(d[0][number]["id"])
			words = p_title.split()
			# If '<=>' does not appears in a post it's not a notification, so process it:
			if not ('<=>' in p_title):
				for word in words:
					if not ('#' in word):
						hashtag = 'nothing (no hashtag included, oops!) Please try again.'
					if ('#' in word):
						if word != ('#' + tag):
							votesmade = True
							hashtag = word
							votes.append(hashtag)
							if not (hashtag in polloptions): 
								hashtag += ', which is not a candidate. Please try again.'
		if votesmade:
			# Save the post number to a file:
			if not (postnum in y):
				f.write(str(postnum) + '\n')
			# Create a poll entry:
			posttext += '• @' + user + ' voted for ' + hashtag + '\n'
	except IndexError:
		pass
	number -= 1
f.close()

# Tidy the message if no-one voted yet:
if not votesmade:
	posttext += '• No-one has voted yet. Why not be first?! Simply create a public post using #' + tag + ' and your choice, not forgetting the hash.\n'

# Total votes:
if votesmade:
	posttext += '\nTotal votes for each option:\n'
	for vote in votes:
		if vote in polloptions:
			polloptions[vote] += 1
	for vote in polloptions:
		posttext += '• ' + str(polloptions[vote]) + ' ' + vote + '\n'	

# Summarise the intermediate result:
if votesmade:
	winner = (max(polloptions.items(), key=operator.itemgetter(1))[0])
	winnervotes = (max(polloptions.items(), key=operator.itemgetter(1))[1])
	posttext += '\nIf the poll was closed now, ' + winner + ' would win with ' + str(winnervotes) + ' votes!'

# FOR TESTING, uncomment the lines in this section to prevent the creation of posts & messages:
# print(posttext)
# posttext = ''

# If there's text to post:
if posttext:
	# Create message in channel 962, using ONLY the text from pnut_message:
	messagecontent = pnutpy.api.create_message(channelid, data={'text': posttext})
	
	# Create a public post:
	pollalert = 'To see the current votes in the TEST #' + tag + ' poll go to the Patter room link below. (Report not run automatically.)'
	channelurl = "https://patter.chat/room/" + channelid
	# Removed hash before tag:
	channelurlmd = '[' + tag + ' <=>](' + channelurl + ")"
	pollalert += "\n" + channelurlmd
	postcontent = pnutpy.api.create_post(data={'text': pollalert})


thmigctrl.py
#!/usr/bin/env python3

# thmigctrl
# v0.1.0 for Python 3.5

# Runs modules based on a matching hashtag

# Setup tag and channel parameters, a list of valid command options: 
tag = 'tmctrl'
retrievecount = 20
channelid = 962

# Import @33MHz and @thrrgilag's library for interacting with pnut.io:
import pnutpy

# Import time, used to delay posting to avoid rate limits:
import time

# Setup pnut.io authorisation:
tokenfile = open("pnut_app_token.txt", "r")
token = tokenfile.read()
token = token.strip()
pnutpy.api.add_authorization_token(token)

# Get hashtag content from pnut.io:
d = pnutpy.api.posts_with_hashtag(tag, count = retrievecount)

# Extract posts, strip out unnecessary words, check for matches to poll options, and construct a summary message:
# Open the previous post numbers file:
f=open('pollctrl.txt','r')
y = f.readlines()
f.close()
f=open('pollctrl.txt','w')
posttext = ''
number = retrievecount
# hashtag = ''
while number >= 0:
	try:
		if not 'is_deleted' in d[0][number]:
			user = str(d[0][number]["user"]["username"])
			querypost = d[0][number]["content"]["text"]
			postnum = str(d[0][number]["id"])
			# If postnum does not appear in the file it's not been seen, so process it to see if a command was made:
			success = False
			if (not (postnum + '\n') in y):
				if 'help' in querypost:
					posttext = '''
*Checks only every 15 minutes.
*Precede all commands with a hash
tmctrl help:
	this!
tmask #hashtag:
	Suggest a hashtag
tmpoll #hashtag:
	Vote for a hashtag
'''
					success = True
				elif ('ask' in querypost):
					posttext = ' ask'
					success = True
				elif ('poll' in querypost):
					posttext = ' poll'
					success = True
				elif success == False:
					posttext = ' Oops, I don\'t understand; please try again. Try #help for more.'
				posttext = '@' + user + posttext + ' (' + postnum + ')'
				if posttext:
					pnutpy.api.create_post(data={'reply_to': postnum, 'text': posttext})
					# Delay to avoid rate limits:
					time.sleep(3.2)
			f.write(str(postnum) + '\n')
			posttext = ''
	except IndexError:
		pass
	number -= 1
f.close()
